packages/file_selector/file_selector_windows/lib/src/file_selector_dart/file_dialog_controller.dart

```diff
@@ -0,0 +1,78 @@
+
+/// A thin wrapper for IFileDialog to allow for faking and inspection in tests.
+///
+/// Since this class defines the end of what can be unit tested, it should
+/// contain as little logic as possible.
+abstract class FileDialogController {
+  /// Creates a controller managing [IFileDialog](https://pub.dev/documentation/win32/latest/winrt/IFileDialog-class.html).
+  FileDialogController(IFileDialog fileDialog) : _fileDialog = fileDialog;
+
+  /// A reference to the file dialog.
+  final IFileDialog _fileDialog;
+
+  /// Disallow copy and assign.
+  //  FileDialogController(const FileDialogController&) = delete;
+  // TODO(eugerossetto): Review implementation of this.
+  @Deprecated('Operation not allowed')
+  FileDialogController operator = (FileDialogController fileDialogController) {
+    throw Exception('Not implemented exception');
+  } as FileDialogController;
+
+  /// Sets the default folder for the dialog to [path].
+  HRESULT setFolder(Pointer<COMObject> path) {
+    return _fileDialog.setFolder(path) as HRESULT;
+  }
+
+  /// Sets the file [name] that is initially shown in the IFileDialog.
+  HRESULT setFileName(String name) {
+    return _fileDialog.setFileName(TEXT(name)) as HRESULT;
+  }
+
+  /// Sets the allowed file type extensions in the IFileOpenDialog.
+  HRESULT setFileTypes(int count, Pointer<COMDLG_FILTERSPEC> filters) {
+    return _fileDialog.setFileTypes(count, filters) as HRESULT;
+  }
+
+  /// Sets the label of the confirmation button.
+  HRESULT setOkButtonLabel(String text) {
+    return _fileDialog.setOkButtonLabel(TEXT(text)) as HRESULT;
+  }
+
+  /// Returns the HRESULT of getting the IFileDialog's [options](https://pub.dev/documentation/win32/latest/winrt/FILEOPENDIALOGOPTIONS-class.html).
+  /// It also uses the [outOptions] paramater to return the current options.
+  HRESULT getOptions(Pointer<Uint32> outOptions) {
+    return _fileDialog.getOptions(outOptions) as HRESULT;
+  }
+
+  /// Sets the [options](https://pub.dev/documentation/win32/latest/winrt/FILEOPENDIALOGOPTIONS-class.html)
+  /// given into an IFileDialog.
+  HRESULT setOptions(int options) {
+    return _fileDialog.setOptions(options) as HRESULT;
+  }
+
+  /// Shows an IFileDialog using the given parent.
+  HRESULT show(int parent) {
+    return _fileDialog.show(parent) as HRESULT;
+  }
+
+  /// Return results from an IFileDialog. This should be used when selecting
+  /// single items.
+  HRESULT getResult(Pointer<Pointer<COMObject>> outItem) {
+    return _fileDialog.getResult(outItem) as HRESULT;
+  }
+
+  /// Return results from an IFileOpenDialog. This should be used when selecting
+  /// single items. This function will fail if the IFileDialog* provided to the
+  /// constructor was not an IFileOpenDialog instance.
+  HRESULT getResults(Pointer<Pointer<COMObject>> outItems) {
+    // TODO(eugerossetto): Review if a free and release [is needed](https://pub.dev/documentation/win32/latest/topics/com-topic.html).
```

Comment:
Reading the documentation we understand that it's needed but, we want to confirm it.

Reply:
I've found [another example](https://learn.microsoft.com/en-us/windows/win32/learnwin32/asking-an-object-for-an-interface) that remarks on the need of releasing the new interface.
```cpp
hr = pFileOpen->QueryInterface(IID_IFileDialogCustomize, 
    reinterpret_cast<void**>(&pCustom));
if (SUCCEEDED(hr))
{
    // Use the interface. (Not shown.)
    // ...

    pCustom->Release();
}
else
{
    // Handle the error.
}
```
Moreover, it's mandatory to verify the operation result.